core/src/main/scala/kafka/log/OffsetIndex.scala

```diff
@@ -114,13 +114,22 @@ class OffsetIndex(file: File, baseOffset: Long, maxIndexSize: Int = -1)
       OffsetPosition(relativeOffset(idx, n), physical(idx, n))
     }
   }
+
+  /**
+    * @param offset that we want to potentially append
+    * @return true if it is safe to append this offset to this index
+    */
+  def canAppend(offset: Long) : Boolean = {
```

Comment:
My reasoning was that if we put the check here, we have to duplicate the same check in TimeIndex as well. So, perhaps we could add this method and the assertion in LogSegment instead.

Also, canAppend() is a bit too general. Perhaps we could rename it to sth like canConvertToRelativeOffset(offset)?